File: daemon/src/engine/canaries.rs
//! Canary (honeypot) sentinel file management.
//!
//! Deploys realistic decoy files into designated user directories. Any write, rename,
//! or deletion event on these sentinel paths serves as an immediate high-fidelity indicator
//! of ransomware activity.

use std::fs;
use std::path::{Path, PathBuf};
use tracing::{info, warn};

/// Directory name used for placing honeypot canary files.
pub const CANARY_DIR_NAME: &str = ".canary_sentinel";

/// Representative decoy file names deployed inside canary directories.
const CANARY_FILES: [&str; 6] = [
    "important_document.docx",
    "financial_report.xlsx",
    "family_photos.zip",
    "passwords.txt",
    "bitcoin_wallet.dat",
    "tax_return_2024.pdf",
];

/// Returns realistic synthetic content for a given canary file name.
fn get_canary_content(filename: &str) -> &'static [u8] {
    match filename {
        "important_document.docx" => {
            b"[SENTINEL_CANARY_DOCX_HEADER]\r\n\
              CONFIDENTIAL CORPORATE STRATEGY & MERGER TARGETS 2024-2026\r\n\
              DO NOT DISTRIBUTE - INTERNAL ONLY\r\n\
              Section 1: Executive Summary\r\n\
              Section 2: Valuation Models and Acquisition Roadmaps\r\n"
        }
        "financial_report.xlsx" => {
            b"[SENTINEL_CANARY_XLSX_HEADER]\r\n\
              Q1-Q4 Audited Balance Sheets, Revenue Forecasting, EBITDA Projections\r\n\
              Account ID: 884-29104-92A\r\n\
              Net Income: $14,892,000\r\n"
        }
        "family_photos.zip" => {
            b"PK\x03\x04\x14\x00\x00\x00\x08\x00\
              SENTINEL_CANARY_ARCHIVE_VACATION_PHOTOS_SUMMER_2023_FAMILY_MEMORIES\
              \x00\x00\x00\x00\x00\x00"
        }
        "passwords.txt" => {
            b"# Sentinel Canary Vault - Personal & Infrastructure Credentials\r\n\
              aws_access_key_id=AKIAIOSFODNN7EXAMPLE\r\n\
              aws_secret_access_key=wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY\r\n\
              root_database_master_pass=S3cur3_C4n4ry_T0k3n_V4ult_99!\r\n"
        }
        "bitcoin_wallet.dat" => {
            b"\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\
              SENTINEL_BITCOIN_CORE_WALLET_DESCRIPTOR_CANARY_v0.21.0\
              \x00\x00\x00\x00\x00\x00"
        }
        "tax_return_2024.pdf" => {
            b"%PDF-1.4\n\
              1 0 obj\n\
              << /Title (Federal Form 1040 - U.S. Individual Income Tax Return 2024) /Author (Sentinel Canary) >>\n\
              endobj\n\
              trailer\n\
              << /Root 1 0 R >>\n\
              %%EOF\n"
        }
        _ => b"[SENTINEL_CANARY_GENERIC_DECOY_FILE]\n",
    }
}

/// Deploys honeypot canary files into `~/.canary_sentinel/` directories across `/home/*` user homes.
///
/// If running as root, also deploys to `/root/.canary_sentinel/`.
///
/// # Returns
/// A vector containing paths to all successfully created canary decoy files.
pub fn deploy_canaries() -> anyhow::Result<Vec<PathBuf>> {
    let mut deployed = Vec::new();
    let mut target_dirs = Vec::new();

    // Enumerate user homes under /home/*
    if let Ok(entries) = fs::read_dir("/home") {
        for entry in entries.flatten() {
            let path = entry.path();
            if path.is_dir() {
                if let Some(name) = path.file_name().and_then(|n| n.to_str()) {
                    if !name.starts_with('.') {
                        target_dirs.push(path.join(CANARY_DIR_NAME));
                    }
                }
            }
        }
    }

    // Include /root if it exists
    let root_home = PathBuf::from("/root");
    if root_home.is_dir() {
        target_dirs.push(root_home.join(CANARY_DIR_NAME));
    }

    for canary_dir in target_dirs {
        match fs::create_dir_all(&canary_dir) {
            Ok(_) => {
                info!("Canary directory initialized at: {}", canary_dir.display());

                // Match ownership to the home directory (e.g. niko:niko) so that user-level
                // ransomware has permission to modify them and trigger the trap.
                let mut target_uid = 0;
                let mut target_gid = 0;
                if let Some(parent) = canary_dir.parent() {
                    if let Ok(meta) = fs::metadata(parent) {
                        use std::os::unix::fs::MetadataExt;
                        target_uid = meta.uid();
                        target_gid = meta.gid();
                    }
                }

                use std::os::unix::ffi::OsStrExt;
                let c_dir = std::ffi::CString::new(canary_dir.as_os_str().as_bytes()).unwrap();
                unsafe { libc::chown(c_dir.as_ptr(), target_uid, target_gid) };

                for file_name in CANARY_FILES {
                    let file_path = canary_dir.join(file_name);
                    let content = get_canary_content(file_name);
                    match fs::write(&file_path, content) {
                        Ok(_) => {
                            let c_file = std::ffi::CString::new(file_path.as_os_str().as_bytes()).unwrap();
                            unsafe { libc::chown(c_file.as_ptr(), target_uid, target_gid) };
                            deployed.push(file_path);
                        }
                        Err(e) => {
                            warn!("Failed to write canary file '{}': {}", file_path.display(), e);
                        }
                    }
                }
            }
            Err(e) => {
                warn!("Failed to create canary directory '{}': {}", canary_dir.display(), e);
            }
        }
    }

    info!("Deployed {} canary files across system honeypots", deployed.len());
    Ok(deployed)
}

/// Checks whether a given file path resides within a canary sentinel directory.
///
/// # Arguments
/// * `path` - Target file path to inspect.
///
/// # Returns
/// `true` if `path` contains [`CANARY_DIR_NAME`].
#[inline]
pub fn is_canary_path(path: &str) -> bool {
    path.contains(CANARY_DIR_NAME)
}

/// Returns a list of all currently existing deployed canary directories.
#[allow(dead_code)]
pub fn get_canary_dirs() -> Vec<PathBuf> {
    let mut dirs = Vec::new();

    if let Ok(entries) = fs::read_dir("/home") {
        for entry in entries.flatten() {
            let path = entry.path();
            if path.is_dir() {
                if let Some(name) = path.file_name().and_then(|n| n.to_str()) {
                    if !name.starts_with('.') {
                        let canary_dir = path.join(CANARY_DIR_NAME);
                        if canary_dir.is_dir() {
                            dirs.push(canary_dir);
                        }
                    }
                }
            }
        }
    }

    let root_canary = PathBuf::from("/root").join(CANARY_DIR_NAME);
    if root_canary.is_dir() {
        dirs.push(root_canary);
    }

    dirs
}

/// Helper function to deploy canaries into a custom base directory (useful for testing and sandboxes).
#[allow(dead_code)]
pub fn deploy_canaries_to_base<P: AsRef<Path>>(base_path: P) -> anyhow::Result<Vec<PathBuf>> {
    let canary_dir = base_path.as_ref().join(CANARY_DIR_NAME);
    fs::create_dir_all(&canary_dir)?;

    let mut deployed = Vec::new();
    for file_name in CANARY_FILES {
        let file_path = canary_dir.join(file_name);
        let content = get_canary_content(file_name);
        fs::write(&file_path, content)?;
        deployed.push(file_path);
    }

    Ok(deployed)
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs;

    #[test]
    fn test_is_canary_path() {
        assert!(is_canary_path("/home/alice/.canary_sentinel/passwords.txt"));
        assert!(is_canary_path("/root/.canary_sentinel/important_document.docx"));
        assert!(is_canary_path(".canary_sentinel/file"));

        assert!(!is_canary_path("/home/alice/Documents/passwords.txt"));
        assert!(!is_canary_path("/home/alice/.config/app.conf"));
        assert!(!is_canary_path("/tmp/canary_sentinel/not_hidden"));
    }

    #[test]
    fn test_canary_files_list_count() {
        assert_eq!(CANARY_FILES.len(), 6);
        for file in CANARY_FILES {
            let content = get_canary_content(file);
            assert!(!content.is_empty(), "Canary content for {} must not be empty", file);
        }
    }

    #[test]
    fn test_deploy_canaries_to_custom_dir() {
        let temp_dir = std::env::temp_dir().join(format!("canary_test_{}", std::process::id()));
        let _ = fs::remove_dir_all(&temp_dir);

        let deployed = deploy_canaries_to_base(&temp_dir).expect("Deploy to custom dir must succeed");
        assert_eq!(deployed.len(), 6);

        for path in &deployed {
            assert!(path.is_file(), "File {} should exist", path.display());
            assert!(is_canary_path(&path.to_string_lossy()));
            let metadata = fs::metadata(path).unwrap();
            assert!(metadata.len() > 0);
        }

        // Cleanup
        let _ = fs::remove_dir_all(&temp_dir);
    }
}

File: src/canaries.c
#include "canaries.h"
#include "sentinel.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <dirent.h>
#include <fcntl.h>
#include <sys/stat.h>
#include <sys/types.h>

static const canary_file_t CANARY_FILES[CANARY_FILE_COUNT] = {
    {
        .name = "important_document.docx",
        .content = "[SENTINEL_CANARY_DOCX_HEADER]\r\n"
                   "CONFIDENTIAL CORPORATE STRATEGY & MERGER TARGETS 2024-2026\r\n"
                   "DO NOT DISTRIBUTE - INTERNAL ONLY\r\n"
                   "Section 1: Executive Summary\r\n"
                   "Section 2: Valuation Models and Acquisition Roadmaps\r\n",
        .size = 197
    },
    {
        .name = "financial_report.xlsx",
        .content = "[SENTINEL_CANARY_XLSX_HEADER]\r\n"
                   "Q1-Q4 Audited Balance Sheets, Revenue Forecasting, EBITDA Projections\r\n"
                   "Account ID: 884-29104-92A\r\n"
                   "Net Income: $14,892,000\r\n",
        .size = 153
    },
    {
        .name = "family_photos.zip",
        .content = "PK\x03\x04\x14\x00\x00\x00\x08\x00"
                   "SENTINEL_CANARY_ARCHIVE_VACATION_PHOTOS_SUMMER_2023_FAMILY_MEMORIES"
                   "\x00\x00\x00\x00\x00\x00",
        .size = 83
    },
    {
        .name = "passwords.txt",
        .content = "# Sentinel Canary Vault - Personal & Infrastructure Credentials\r\n"
                   "aws_access_key_id=AKIAIOSFODNN7EXAMPLE\r\n"
                   "aws_secret_access_key=wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY\r\n"
                   "root_database_master_pass=S3cur3_C4n4ry_T0k3n_V4ult_99!\r\n",
        .size = 208
    },
    {
        .name = "bitcoin_wallet.dat",
        .content = "\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00"
                   "SENTINEL_BITCOIN_CORE_WALLET_DESCRIPTOR_CANARY_v0.21.0"
                   "\x00\x00\x00\x00\x00\x00",
        .size = 72
    },
    {
        .name = "tax_return_2024.pdf",
        .content = "%PDF-1.4\n"
                   "1 0 obj\n"
                   "<< /Title (Federal Form 1040 - U.S. Individual Income Tax Return 2024) /Author (Sentinel Canary) >>\n"
                   "endobj\n"
                   "trailer\n"
                   "<< /Root 1 0 R >>\n"
                   "%%EOF\n",
        .size = 151
    }
};

bool is_canary_path(const char *path) {
    if (path == NULL) {
        return false;
    }
    return strstr(path, CANARY_DIR_NAME) != NULL;
}

static int write_canary_file(const char *dir_path, const canary_file_t *file, uid_t uid, gid_t gid) {
    char file_path[MAX_PATH_LEN];
    int n = snprintf(file_path, sizeof(file_path), "%s/%s", dir_path, file->name);
    if (n < 0 || (size_t)n >= sizeof(file_path)) {
        return -1;
    }

    int fd = open(file_path, O_WRONLY | O_CREAT | O_TRUNC, 0644);
    if (fd < 0) {
        return -1;
    }

    size_t written = 0;
    while (written < file->size) {
        ssize_t ret = write(fd, file->content + written, file->size - written);
        if (ret <= 0) {
            close(fd);
            return -1;
        }
        written += (size_t)ret;
    }

    close(fd);

    if (uid != 0 || gid != 0) {
        if (chown(file_path, uid, gid) != 0) {
            /* non-fatal if running as non-root */
        }
    }

    return 0;
}

int deploy_canaries_to_base(const char *base_path) {
    if (base_path == NULL) {
        return -1;
    }

    char canary_dir[MAX_PATH_LEN];
    int n = snprintf(canary_dir, sizeof(canary_dir), "%s/%s", base_path, CANARY_DIR_NAME);
    if (n < 0 || (size_t)n >= sizeof(canary_dir)) {
        return -1;
    }

    struct stat st;
    uid_t uid = 0;
    gid_t gid = 0;
    if (stat(base_path, &st) == 0) {
        uid = st.st_uid;
        gid = st.st_gid;
    }

    if (mkdir(canary_dir, 0755) != 0 && stat(canary_dir, &st) != 0) {
        return -1;
    }

    if (uid != 0 || gid != 0) {
        if (chown(canary_dir, uid, gid) != 0) {
            /* non-fatal if running as non-root */
        }
    }

    int deployed = 0;
    for (size_t i = 0; i < CANARY_FILE_COUNT; i++) {
        if (write_canary_file(canary_dir, &CANARY_FILES[i], uid, gid) == 0) {
            deployed++;
        }
    }

    return deployed;
}

int deploy_canaries(void) {
    int total_deployed = 0;

    DIR *dir = opendir("/home");
    if (dir != NULL) {
        struct dirent *entry;
        while ((entry = readdir(dir)) != NULL) {
            if (entry->d_name[0] == '.') {
                continue;
            }

            char user_home[MAX_PATH_LEN];
            int n = snprintf(user_home, sizeof(user_home), "/home/%s", entry->d_name);
            if (n < 0 || (size_t)n >= sizeof(user_home)) {
                continue;
            }

            struct stat st;
            if (stat(user_home, &st) == 0 && S_ISDIR(st.st_mode)) {
                int res = deploy_canaries_to_base(user_home);
                if (res > 0) {
                    total_deployed += res;
                }
            }
        }
        closedir(dir);
    }

    struct stat root_st;
    if (stat("/root", &root_st) == 0 && S_ISDIR(root_st.st_mode)) {
        int res = deploy_canaries_to_base("/root");
        if (res > 0) {
            total_deployed += res;
        }
    }

    return total_deployed;
}
